File: COMP1005/a3/a3.py
import signal
from random import shuffle 

def pose_question(question, answer_a, answer_b, answer_c, answer_d, correct_answer):
	print '\n%s\n A: %s\n B: %s\n C: %s\n D: %s\
			'%(question, answer_a, answer_b, answer_c, answer_d)

	answer = raw_input("Answer: ")
	answered = False
	while not answered:
		if (answer.lower() == "a" or answer.lower() == "b" or answer.lower() == "c" or answer.lower() == "d"):
			answered = True
		else:
			answer = raw_input("(A, B, C or D): ")

	if answer.lower() == correct_answer:
		print "Correct! score++"
		return 1
	else:
		print "False"
		return 0

#shuffles the questions, calls pose_question, and handles endgame state
def play():
	questions = [
	{"question":"[x for x in range(5)] is a:",
	"a":"For loop",
	"b":"While Loop",
	"c":"Magic",
	"d":"List comprehension",
	"answer":"d"
	},
	{"question":"Python is a:",
	"a":"Interpreted Language",
	"b":"Compiled Language",
	"c":"Punchcard Language",
	"d":"Foreign Language",
	"answer":"a"
	},
	{"question":"Where is John's office?",
	"a":"1170 HP",
	"b":"5380 HP",
	"c":"4155 HP",
	"d":"KM-TH SA",
	"answer":"b"
	},
	{"question":"Linux was originally programmed by:",
	"a":"Ken Thompson and Dennis Ritchie",
	"b":"Richard Stallman",
	"c":"Linus Torvalds",
	"d":"DEVELOPERS DEVELOPERS DEVELOPERS DEVELOPERS", #see also: https://www.youtube.com/watch?v=KMU0tzLwhbE
	"answer":"c"
	},
	{"question":"What algorithm was Google founded on?",
	"a":"MapReduce",
	"b":"PageRank",
	"c":"GoogleRank",
	"d":"A* Graph Search",
	"answer":"b"}]
	
	shuffle(questions)
	score = 0
	for q in questions:
		score += pose_question(q["question"], q["a"], q["b"], q["c"], q["d"], q["answer"])

	print "\nscore:", score,
	if score > 5:
		print '"hacker"'
	elif score >= 4:
		print '"10/10 good job!"'
	elif score >= 2:
		print '"It went okay"'
	else:
		print '"awww..."'

	while True:
		again = raw_input("play again? y/n: ")
		if again.lower() == "y":
			play()
		elif again.lower() == "n":
			print "bye!"
			exit()

#lets us exit without spewing a stack trace
signal.signal(signal.SIGINT, lambda s, f: exit())

play()

File: COMP1005/a3/a3.cpp
#include <map>
#include <string>
#include <vector>
#include <chrono>
#include <random>
#include <iostream>
#include <algorithm>

using namespace std;

int pose_question(string question, string answer_a, string answer_b, string answer_c, string answer_d, string correct_answer){
	cout<<"\n"<<question<<"\n A: "<<answer_a<<"\n B: "<<answer_b<<"\n C: "<<answer_c<<"\n D: "<<answer_d<<"\n";
	string answer;
	cout << "Answer: ";
	cin >> answer;
	bool answered = false;
	while (!answered){
		transform(answer.begin(), answer.end(), answer.begin(), ::tolower);
		if (answer == "a" || answer == "b" || answer == "c" || answer == "d"){
			answered = true;
		} else {
			cout << "(A, B, C or D): ";
			cin >> answer;
		}
	}

	if (answer == correct_answer){
		cout << "Correct ++score\n";
		return 1;
	} else {
		cout << "false\n";
		return 0;
	}
}

int play(){
	vector<map<string, string>> questions = {
		{
			{"question", "[x for x in range(5)] is a:"},
			{"a","For loop"},
			{"b","While Loop"},
			{"c","Magic"},
			{"d","List comprehension"},
			{"answer", "d"}
		},{
			{"question", "Python is a:"},
			{"a","Interpreted Language"},
			{"b","Compiled Language"},
			{"c","Punchcard Language"},
			{"d","Foreign Language"},
			{"answer", "a"}
		},{
			{"question", "Where is John's office?"},
			{"a","1170 HP"},
			{"b","5380 HP"},
			{"c","4155 HP"},
			{"d","KM-TH SA"},
			{"answer", "b"}
		},{
			{"question", "Linux was originally programmed by:"},
			{"a","Ken Thompson and Dennis Ritchie"},
			{"b","Richard Stallman"},
			{"c","Linus Torvalds"},
			{"d","DEVELOPERS DEVELOPERS DEVELOPERS DEVELOPERS"},
			{"answer", "c"}
		},{
			{"question", "What algorithm was Google founded on?"},
			{"a","MapReduce"},
			{"b","PageRank"},
			{"c","GoogleRank"},
			{"d","A* Graph Search"},
			{"answer", "b"}
		},
	};
	unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
	shuffle(questions.begin(), questions.end(), default_random_engine(seed));
	int score = 0;
	for_each(questions.begin(), questions.end(), [questions, &score] (map<string, string> q) {
		score += pose_question(q["question"], q["a"], q["b"], q["c"], q["d"], q["answer"]);
	});

	cout << "\nscore: " << score << "\n";
	if (score > 5){
		cout << "hacker";
	} else if (score >= 4){
		cout << "10/10 good job!";
	} else if (score >= 2){
		cout << "It went okay";
	} else {
		cout << "awww...";
	}

	while (1){
		string again;
		cout << "\nplay again? y/n:";
		cin >> again;
		transform(again.begin(), again.end(), again.begin(), ::tolower);
		if (again == "y"){
			play();
		} else if (again == "n"){
			cout << "bye!\n";
			return 0;
		}
	}
}

int main(int argc, char const *argv[]){
	//[]{cout << "Lambda!";}();
	play();
	return 0;
}